### ISSUE TITLE
Destructuring assignment

### ISSUE THREAD

**Issue by munificent** [dart-lang/language#2438]
The [patterns proposal](https://github.com/dart-lang/language/blob/master/working/0546-patterns/patterns-feature-specification.md) currently supports destructuring *variable declarations*, but it has no way to use destructuring to assign to existing variables. Should it?

I think the answer is generally "yes", though it raises interesting questions about: 

* What kinds of patterns are allowed on the left-hand side of an `=`? Certainly not refutable ones. Can you also declare variables?

* Where a destructuring assignment expression is allowed to appear? Do we only allow this for assignments in "statement position" inside an expression statements, or anywhere an assignment can occur? If it declares variables, what is their scope?

* Are there any ambiguities with existing assignment target syntax?

* Does it work with any compound assignment operators? Probably not.

### Syntax

We extend `expression`:

```
expression  ::= patternAssignment
              | // Existing productions...

patternAssignment ::= pattern '=' expression
```

As with pattern variable declarations, we restrict what kind of pattern the outermost pattern can be. It must be one of:

* groupingPattern
* listPattern
* mapPattern
* recordPattern
* extractorPattern
* nullCheckPattern

Also:

* An assignment is an irrefutable context, so it's a compile error to have a refutable pattern in there.

* It's a compile-time error if a variable pattern has a type annotation.

Note that this syntax allows chaining pattern assignments (probably not very useful) but doesn't allow any compound assignment operators.

### Semantics

All variable and cast patterns are interpreted as assigning to an existing variable (or setter) with that name using the normal scope resolution.

You can't declare any new variables. Disallowing this allows pattern assignment expressions to appear anywhere expressions are allowed without raising tricky questions around the scope of any new variables.

It's a compile time error if a variable or cast pattern refers to a name that can't be resolved. In assignments, variable and cast patterns are always treated as irrefutable, and their static type is the type of the assigned variable or setter.

The runtime semantics are a little subtle because patterns (even irrefutable ones) can have side effects, as can assignment. Consider:

```dart
var a = 'before';
try {
  Object o = 'not int';
  [a, b as int] = ['after', o];
} catch (_) {
  // Ignore exception.
}
print(a);
```

Should this print "before" or "after"?

An even stranger one is:

```dart
var x = 'x';

class FunnyList implements List<String> {
  String operator[](int index) => '$x!';

  int get length => 2;
  // Other List methods...
}

main() {
  var y = 'y before';
  [x, y] = FunnyList();
  print('$x $y');
}
```

Does this print "x! x!!" or "x! x!"?

I think the right thing is to defer all assignments until after the destructuring is done, so:

1.   Evaluate the right-hand side expression to a value `v`.
2.   Match `v` against the pattern on the left. When matching a variable or cast pattern against a value `o`, record that `o` will be the new value for the corresponding variable/setter, but do not store the variable or invoke the setter.
3.   Once all destructuring and matching is done, store all of the assigned variables and invoke the setters with their corresponding values.

In other words, it's as if every variable or cast pattern in an assignment expression *is* a new variable declaration (with a hidden name). And then after the assignment expression and matching completes, those temporary variables are all then written to the corresponding real variables and setters.

**Comment by lrhn** [comment]
I agree with the parallel assignment rule: Evaluate RHS, perform all operations in the LHS patterns, then finally assign the results.

For `int x,y; (x!, y!) = (42, null as int?);` that means doing the `null!` before assigning to `x`.  (The `!` is the null-check pattern, right?)

**Comment by munificent** [comment]
> I'm guessing you mean "nullAssertPattern" instead of "nullCheckPattern", since the latter, `pattern?`, is refutable.

That's the list of patterns that are allowed to appear syntactically as the outermost pattern in a `patternDeclaration`. That includes the refutable `nullCheckPattern` because `patternDeclaration` is also used by pattern-if statements. It enables:

```dart
if (var foo? = someNullableThing()) { ... }
```

But if you use a null check pattern in a variable declaration or destructuring assignment, it's a compile error because you can't have a refutable pattern in an irrefutable context. So the end result is that only the right patterns are allowed, but the reason this particular pattern is disallowed is a little subtle.

**Comment by munificent** [comment]
> If we did allow assignment to the null-check pattern, `x? = e`, as a _refutable assignment_, it's actually potentially useful

Refutable conditional assignment is a neat idea, but I think we'd want some explicit syntax to opt in to that instead of just making a refutable pattern on the left spontaneously change the behavior of the assignment.
